framework/src/postprocessors/PseudoTimestep.C

```diff
@@ -0,0 +1,220 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// Navier-Stokes includes
+#include "PseudoTimestep.h"
+#include "FEProblemBase.h"
+#include "NonlinearSystemBase.h"
+#include "Transient.h"
+#include "Restartable.h"
+
+// MOOSE includes
+#include "libmesh/quadrature.h"
+#include "libmesh/enum_norm_type.h"
+#include "metaphysicl/raw_type.h"
+
+#include <algorithm>
+#include <limits>
+
+registerMooseObject("MooseApp", PseudoTimestep);
+
+InputParameters
+PseudoTimestep::validParams()
+{
+  InputParameters params = GeneralPostprocessor::validParams();
+
+  /// Enum that is used to select the timestep-selection method.
+  MooseEnum method("SER=1 EXP=2 RDM=3");
+  method.addDocumentation("SER", "Switched evolution relaxation");
+  method.addDocumentation("EXP", "Exponential progression");
+  method.addDocumentation("RDM", "Residual difference method");
+  params.addParam<MooseEnum>("method", method, "The method used for CFL scaling");
+  params.addRequiredParam<Real>("initial_dt", "Initial timestep");
+  params.addRequiredParam<Real>("alpha", "The parameter alpha used in the scaling of the timestep");
+  params.addParam<Real>("max_dt", "The largest timestep allowed");
+  params.addParam<unsigned int>(
+      "iterations_window",
+      "For how many iterations should the CFL be tracked (only applies to the SER method)");
+  // Because this post-processor is meant to be used with PostprocessorDT, it
+  // should be executed on initial (not included by default) and timestep end.
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_TIMESTEP_END, EXEC_INITIAL};
+  params.suppressParameter<ExecFlagEnum>("execute_on");
+
+  params.addClassDescription("Computes pseudo-time steps for obtaining steady-state solutions "
+                             "through a pseudo transient process.");
+
+  return params;
+}
+
+PseudoTimestep::PseudoTimestep(const InputParameters & parameters)
+  : GeneralPostprocessor(parameters),
+    _method(getParam<MooseEnum>("method")),
+    _initial_dt(getParam<Real>("initial_dt")),
+    _alpha(getParam<Real>("alpha")),
+    _residual_norms_sequence(declareRestartableData<std::vector<Real>>("residual_norms_sequence")),
+    _iterations_step_sequence(declareRestartableData<std::vector<Real>>("iterations_step_sequence"))
+{
+  if (!_fe_problem.isTransient())
+    mooseError("This pseudotimestepper can only be used if the steady state is computed via time "
+               "integration.");
+
+  if (parameters.isParamValid("max_dt"))
+  {
+    _max_dt = getParam<Real>("max_dt");
+    _bound = true;
+  }
+
+  if (_method == 1)
+  {
+    if (parameters.isParamSetByUser("iterations_window"))
+      _iterations_window = getParam<unsigned int>("iterations_window");
+    else
+      _iterations_window = 1;
+  }
+  else if (parameters.isParamSetByUser("iterations_window"))
```

Comment:
```suggestion
  else if (parameters.isParamValid("iterations_window"))
```

Reply:
isParamSetByUser wont catch someone passing it wrong programmatically in an Action

Reply:
paramValid doesn't work if you want me to throw an error, which is what I changed it from.